Fix first 2 issues of DAN-170: Added ISIS topology and mpls-te configuration

hariosniral opened the pull request:
1. It is the fix for the first 2 issues of DAN-170
2. Support added for topology and mpls-te in danos.


Signed-off-by: harios_niral <hari@niralnetworks.com>

deastoe committed:
BGP: Add ebgp-requires-policy global parameter

Since FRR 7.4, by default, import and export policies must be
specified for EBGP peers (RFC 8212) [1]. In earlier versions (those
used in DANOS) this behaviour was not enabled by default, but could
be enabled by the "bgp ebgp-requires-policy" (FRR) config.

To allow this behaviour change to be reverted via DANOS configuration
"bgp ebgp-requires-policy" is now represented by the
ebgp-requires-policy leaf under the BGP instance parameters
container(s).

The default value is "enabled" which represents the default upstream
behaviour from FRR 7.4. Therefore using this YANG module with earlier
FRR versions will also result in this behaviour being configured.

The presence statement is removed from the BGP instance parameters
containers to expose this default value whenever BGP configuration is
applied.

VRVDR-51589

[1] https://github.com/FRRouting/frr/releases/tag/frr-7.4

deastoe committed:
BGP: Add "network import-check" global parameter

Since FRR 7.4, by default, bgpd will not advertise networks unless
they are present in the RIB [1]. In earlier versions (those used in
DANOS) this behaviour was not enabled by default, but could be
enabled by the "bgp network import-check" (FRR) config.

To allow this behaviour change to be reverted via DANOS configuration
"bgp network import-check" is now represented by the import-check leaf
under the network child container of the BGP instance parameters
container(s).

The default value is "enabled" which represents the default upstream
behaviour from FRR 7.4. Therefore using this YANG module with earlier
FRR versions will also result in this behaviour being configured.

VRVDR-51589

[1] https://github.com/FRRouting/frr/pull/6228

deastoe committed:
parser.py: do subst. on enter/exit translations

Use common command processing for regular node translations as well
as the special case node enter/exit translations.

This makes the value substitution capabilities, and ability to specify
a list of translations for a single path, available to the enter/exit
translations.

deastoe committed:
Add common translations for interface config

Several different protocols define configuration which is translated
to interface mode commands. Interface mode is entered by issuing the
"interface <name>" command and it makes sense to issue this only once
for each interface.

Therefore we define a new translation file to create the interface
mode config, and remove the duplicate config from the various protocol
specific translation files.

This also provides a single place where an exit translation could be
defined for interfaces.

deastoe committed:
parser.py: re-factor json imports

Import json module rather than only load() and loads(). This requires
using the json namespace to call these functions, which makes the
code a little clearer.

deastoe committed:
parser.py: Add SyntaxFile class

SyntaxFile represents a single syntax (translation) file on disk. A
load() method is implemented which returns the syntax dictionary as
defined by the JSON object in the file.

VyattaJSONParser.discover_syntax() generates a dictionary of SyntaxFile
objects, keyed by filename, and stored as the syntax_files instance
variable.

VyattaJSONParser.read_syntax_files() is updated to work on SyntaxFile
objects, after a call to discover_syntax().

hariosniral committed:
Fix first 2 issues of DAN-170: Added ISIS topology and mpls-te configuration

1. It is the fix for the first 2 issues of DAN-170.
2. Support added for topology and mpls-te in danos.

Signed-off-by: harios_niral <hari@niralnetworks.com>

deastoe committed:
parser.py: prioritise config before reading syntax

Prioritising the input configuration does not require the syntax
(translation) files to be loaded. Therefore perform prioritisation
immediately after loading the input config.

In future this change will allow the configuration parsing to be more
finely tuned by allowing only a subset of syntax files to be loaded
at any one time, effectively determining which parts of the input
configuration are translated.

We also make a change to the VyattaJSONParser constructor to ensure
that the syntax dictionary is always (empty) allocated. This is
required since it is referenced in the prioritize() call stack:

  depth_first_traverse()
    on_exit()
      retrieve_commands()
        syntax.get()

deastoe committed:
parser.py: Add steps support

This commit allows the parsing to be broken down into smaller logical
chunks through the new steps.json configuration file.

steps.json defines a JSON array of objects with each object
representing a parsing "step". To complete the parsing operation
steps are executed sequentially as defined in steps.json.

The steps.json configuration as defined in this commit results in the
same behaviour as before.

See the additions to scripts/frr/README.md in this commit for further
information.

deastoe committed:
parser.py: Replace base.conf with step config

Replace the hardcoded handling of the static base.conf configuration
file with static configuration applied by a parsing step.

Each parsing step can apply arbitrary static lines to the output
configuration, which makes a great replacement for base.conf.

deastoe committed:
steps.json: Apply static translations first

Previously in cases where static route configuration is applied in a
routing instance along with BGP (in default or non-default routing
instance) frr-reload.py could fail as the "vrf" line (of the
static route configuration) was interpreted as the vrf-policy command
of BGP mode.

The root cause is somewhat multi-faceted. The frr.conf file we
generate as input to frr-reload.py triggers a failure in the
`vtysh -m` processing (spawned by frr-reload.py), since due to the
above we attempt to execute "exit-vrf" when not in vrf mode, which
yields an unknown command error. See [1].

This can be addressed by updating the BGP translations to ensure that
each "router bgp" configuration block is terminated with an "exit"
line - `vtysh -m` will now happily mark the file.

However the issue still remains as frr-reload.py, following its
processing, does not retain these "exit" lines. As a result when it
goes to apply the configuration changes via `vtysh -f` it still
attempts to enter the "vrf" line of the static route configuration
while in BGP mode.

Since this breakage is caused by the ordering of the input config
to frr-reload.py we can workaround it by simply moving all static
route configuration to the start of the input config.

Having the static route configuration output prior to any protocol
configuration does also match the behaviour from vtysh "show run".

DAN-176

[1] https://github.com/FRRouting/frr/issues/7326

deastoe committed:
priorities.json: Fix BGP remote-as in rt. instance

The remote-as neighbour/peer-group configuration must be applied
before any other configuration for the same neighbour/peer-group.

This already worked as expected for neighbors/peer-groups defined
in the default routing instance, but not for those defined in other
routing instances.

DAN-216

deastoe committed:
1.14.0

KaushikNiral commented:
In the same PR we will add the fix for the 3rd issue also mentioned in https://danosproject.atlassian.net/browse/DAN-170

deastoe commented on scripts/frr/configs/commands/isis.json:
```diff
@@ -12,6 +12,13 @@
     "/interfaces/@element/isis/network": "isis network {/@text}",
     "/interfaces/@element/isis/passive": "isis passive",
+    "/interfaces/@element/isis/topology/ipv4-mgnt": "isis topology ipv4-mgnt",
+    "/interfaces/@element/isis/topology/ipv4-multicas": "isis topology ipv4-multicas",
```
`multicas` typos.

hariosniral replied:
Resolved.

deastoe commented on scripts/frr/configs/commands/isis.json:
```diff
@@ -12,6 +12,13 @@
     "/interfaces/@element/isis/password/md5": "isis password md5 {/@text}",
     "/interfaces/@element/isis/network": "isis network {/@text}",
     "/interfaces/@element/isis/passive": "isis passive",
+    "/interfaces/@element/isis/topology/ipv4-mgnt": "isis topology ipv4-mgnt",
```
`mgnt` typos (and for ipv6 below).

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
+      }
+
+      leaf inter-as {
+	type level;
```
Please use space indent.

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
+    container mpls-te {
+      configd:help "MPLS-TE specific commands.";
+      leaf router-address {
+	type types:ipv4-address;
```
Please use space indent.

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
+        description "Enable the MPLS-TE functionality.";
+      }
+
+      leaf inter-as {
```
AFAICS this configuration is not yet implemented: https://github.com/FRRouting/frr/blob/93ca501b61d924b662712ae4fc9273782e8f0396/isisd/isis_cli.c#L1221

This configuration should not be exposed by DANOS until it is supported in FRR.

hariosniral replied:
The leaf has been removed.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
     }
+
+    container mpls-te {
+      configd:help "MPLS-TE specific commands.";
```
Instead of having an "on" leaf, I think it might be cleaner to make this a presence container. Making a container have presence essentially means that it has some semantic meaning, and may be configured without configuring any of its children [1].

You can use something like this to make this a presence container:
```
presence "Enable MPLS-TE";
```
With a translation like:
```
"/protocols/isis/@element/mpls-te/@enter": "mpls-te on",
```
Then whenever this container is configured (with or without children) the functionality will be enabled.

[1] https://tools.ietf.org/html/rfc6020#section-7.5.1

hariosniral replied:
The solution given by you is handled.
Thanks for review.

deastoe replied:
Looks good, thanks.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
+      configd:help "Configure IS-IS topologies.";
+      description "Configure IS-IS topologies.";
+      container ipv4-multicast {
+	presence "Present if a separate IPv4-multicast topology is configured for this area.";
```
Inconsistent indent on this and some of the below presence statements.

deastoe replied:
Also suggest dropping `Present if` from these descriptions.

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -561,6 +670,57 @@ module vyatta-protocols-frr-isis-v1 {
+        leaf ipv4-multicast {
+          type empty;
+          description
+	    "IPv4 multicast topology.";
```
Inconsistent indent.

hariosniral replied:
Resolved.

deastoe replied:
This one still looks off.

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -561,6 +670,57 @@ module vyatta-protocols-frr-isis-v1 {
+          configd:help "IPv4 multicast topology.";
+        }
+
+        leaf ipv4-mgnt {
```
Should use `mgmt` for consistency (same for ipv6).

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -396,6 +403,108 @@ module vyatta-protocols-frr-isis-v1 {
+	presence "Present if a separate IPv4-multicast topology is configured for this area.";
+        description
+          "IPv4 multicast topology.";
+        configd:help"IPv4 multicast topology.";
```
Please add a space after `configd:help`.

hariosniral replied:
Resolved.

deastoe committed:
bgp.json: Improve peer-group translations

The neighbor should be assigned to a peer-group in bgp mode rather
than under each address family.

When the peer-group is assigned under an address family it is
represented in the FRR running configuration as though it were
configured in the parent bgp mode.

Therefore frr-reload.py determines a diff between the running and
desired configuration which results in un-necessary re-configuration,
and therefore peer flaps.

With the new translations we generate the peer-group assignment
configuration in bgp mode regardless of what address family it was
configured under (we first try ipv4-unicast, then ipv6-unicast).
This is OK since validation ensures that the same peer-group is
assigned for each address family of a given neighbor.

In this commit we also remove some unused (they expect a "neighbor"
node beneath each address family, which does not exist) translations
from the peer-group configuration trees.

DAN-212

deastoe committed:
1.14.1

deastoe committed:
vyatta-frr-vci.postinst expects frr.conf

The vyatta-frr-vci.postinst script expected a desired frr.conf file
to exist in /etc/vyatta-routing so it could generate a default
/etc/frr/frr.conf.

Since a96e4563e3953992d26dc5e3202ff956c7c5a100 there is no longer
an /etc/vyatta-routing/frr.conf on package install.

Therefore use parser.py to generate the default /etc/frr/frr.conf.
A "--no-reload" option is added to parser.py to prevent it calling
frr-reload.py. The postinst script uses this option to maintain
its previous behaviour.

deastoe committed:
debian: Drop dh-exec dependency

No longer required.

deastoe committed:
1.14.2

Commit:
VRVDR-53646 Add interface ipv6 isis instance CLI

deastoe commented on scripts/frr/configs/commands/isis.json:
```diff
@@ -12,6 +12,13 @@
     "/interfaces/@element/isis/password/md5": "isis password md5 {/@text}",
     "/interfaces/@element/isis/network": "isis network {/@text}",
     "/interfaces/@element/isis/passive": "isis passive",
+    "/interfaces/@element/isis/topology/ipv4-mgmt": "isis topology ipv4-mgnt",
```
`mgnt` typo in the vtysh command.

hariosniral replied:
Resolved.

deastoe commented on scripts/frr/configs/commands/isis.json:
```diff
@@ -12,6 +12,13 @@
+    "/interfaces/@element/isis/topology/ipv4-multicast": "isis topology ipv4-multicas",
+    "/interfaces/@element/isis/topology/ipv4-unicast": "isis topology ipv4-unicast",
+    "/interfaces/@element/isis/topology/ipv6-dstsrc": "isis topology ipv6-dstsrc",
+    "/interfaces/@element/isis/topology/ipv6-mgmt": "isis topology ipv6-mgnt",
```
`mgnt` typo in the vtysh command.

hariosniral replied:
Resolved.

deastoe commented on yang/vyatta-protocols-frr-isis-v1.yang:
```diff
@@ -5,6 +5,9 @@ module vyatta-protocols-frr-isis-v1 {
     prefix protocols;
   }
+  import vyatta-types-v1 {
+        prefix types;
```
Please use consistent indentation.

hariosniral replied:
Resolved.

deastoe commented on scripts/frr/configs/commands/isis.json:
```diff
@@ -12,6 +12,13 @@
     "/interfaces/@element/isis/network": "isis network {/@text}",
     "/interfaces/@element/isis/passive": "isis passive",
+    "/interfaces/@element/isis/topology/ipv4-mgmt": "isis topology ipv4-mgnt",
+    "/interfaces/@element/isis/topology/ipv4-multicast": "isis topology ipv4-multicas",
```
It's functionally fine, but it would be better to list the vtysh command. `ipv4-multicas` -> `ipv4-multicast`.

hariosniral replied:
Resolved.

deastoe committed:
BGP: import-check & ebgp-requires-policy help text

The recently added import-check and ebgp-requires-policy parameter
enums were missing help and description text - add it.

VRVDR-51589

deastoe commented:
Hi @hariosniral,

There is now a conflict in these diffs - please could you do a rebase.

To resolve the revision statement conflict, please simply update the description on the existing "2020-11-17" revision (no need to add a new revision statement).

Please also squash the commits so there is one commit with the new topology and mpls-te configuration, and another commit with just the level-1/level-2 fix.

Thanks.

deastoe committed:
1.14.3

hariosniral commented:
> Hi @hariosniral,
> 
> There is now a conflict in these diffs - please could you do a rebase.
> 
> To resolve the revision statement conflict, please simply update the description on the existing "2020-11-17" revision (no need to add a new revision statement).
> 
> Please also squash the commits so there is one commit with the new topology and mpls-te configuration, and another commit with just the level-1/level-2 fix.
> 
> Thanks.

Hi @deastoe,
Please follow PR #9 . In this PR did something wrong, so raised new PR #9 .
